contracts/Multi-signature/ApprovalCollector.sol

```diff
@@ -0,0 +1,215 @@
+        }
+
+        emit ApprovalReceived(msg.sender, transactionId);
+
+        if (_transactions[transactionId].classParams.neededApprovals == 0 && _transactions[transactionId].classParams.neededCriticalApprovals == 0) {
+            emit AllApprovalsReceived(transactionId);
+            _removePendingTransaction(transactionId);
+            _transactions[transactionId].timestamp = 0;
+            multiplexingContract.performTransaction(_transactions[transactionId].executor, _transactions[transactionId].transaction);
+        }
+    }
+
+    /**
+     * @dev Checks is account approved transaction
+     * @param approver Address of account
+     * @dev approver must have permission for approving
+     * @param transactionId Identifier of transaction
+     * @return true if approver approved transaction with transactionId, and false if not
+     */
+    function hasApproved(address approver, bytes32 transactionId) public view onlyAdmin returns (bool) {
+        return _hasApproved(approver, transactionId);
+    }
+
+    /**
+     * @dev Function for obtaining pending transactions
+     * @return array with pending transaction identifiers
+     */
+    function getPendingTransactions() public view returns (bytes32[] memory) {
+        return _pendingTransactions;
+    }
+
+    /**
+     * @dev Function for obtaining pending transaction details
+     * @param transactionId Identifier of transaction
+     * @return address of contract, that should execute transaction and encoded function selector and parameters
+     */
+    function getTransactionInfo(bytes32 transactionId) public view returns (address, bytes memory, uint, uint) {
+        return (
+            _transactions[transactionId].executor,
+            _transactions[transactionId].transaction,
+            _transactions[transactionId].classParams.neededApprovals,
+            _transactions[transactionId].classParams.neededCriticalApprovals
+        );
+    }
+
+    function addAdministrator(address newAdmin) public {
+        require(_criticalApprovers[msg.sender]);
+        _administrators[newAdmin] = true;
+    }
+
+    function deleteAdministrator(address approver) public {
+        require(_criticalApprovers[msg.sender]);
+        _administrators[approver] = false;
+    }
+
+    function addCriticalApprover(address newApprover) public onlyOwner {
+        _criticalApprovers[newApprover] = true;
+    }
+
+    function deleteCriticalApprover(address approver) public onlyOwner {
+        _administrators[approver] = false;
+    }
+
+    function setContractClass(address _contract, ContractClass class) public onlyAdmin {
+        _contractClass[_contract] = class;
+    }
+
+    /**
+     * @dev Function for getting current controlled contract
+     */
+    function getMultiplexingContract() public view returns (MultiplexingContract) {
+        return multiplexingContract;
+    }
+
+    function updateMultiplexingContract(MultiplexingContract newContractAddress) public onlyOwner {
```

Comment:
This should not be possible at all, or at least require a vote. Otherwise, Admins could vote on one transaction to be executed, and then any admin (even the lower ones) could replace the contract being executed.